plugins/snowflake/dbt/include/snowflake/macros/materializations/incremental.sql

```diff
@@ -0,0 +1,75 @@
+
+{% materialization incremental, adapter='snowflake' -%}
+
+  {%- set unique_key = config.get('unique_key') -%}
+  {%- set sql_where = config.get('sql_where') -%}
+  {%- set non_destructive_mode = (flags.NON_DESTRUCTIVE == True) -%}
+  {%- set full_refresh_mode = (flags.FULL_REFRESH == True) -%}
+  {%- set identifier = model['alias'] -%}
+
+  {%- set old_relation = adapter.get_relation(database=database, schema=schema, identifier=identifier) -%}
+  {%- set target_relation = api.Relation.create(database=database, identifier=identifier, schema=schema, type='table') -%}
+  {%- set exists_as_table = (old_relation is not none and old_relation.is_table) -%}
+  {%- set exists_not_as_table = (old_relation is not none and not old_relation.is_table) -%}
+  {%- set force_create = full_refresh_mode -%}
+
+  -- setup
+
+  {% set source_sql -%}
+     -- wrap sql in parens to make it a subquery --
+     (
+        select * from (
+            {{ sql }}
+        )
+    )
+  {%- endset -%}
+
+  {{ run_hooks(pre_hooks, inside_transaction=False) }}
+
+  -- `BEGIN` happens here:
+  {{ run_hooks(pre_hooks, inside_transaction=True) }}
+
+  -- build model
+  {% if force_create or old_relation is none -%}
+    {%- call statement('main') -%}
+
+      {# -- create or replace logic because we're in a full refresh or table is non existant. #}
+      {% if old_relation is not none and old_relation.type == 'view' %}
+        {# -- I'm preserving one of the old checks here for a view, and to make sure Snowflake doesn't
+        -- complain that we're running a replace table on a view. #}
+          {{ log("Dropping relation " ~ old_relation ~ " because it is a view and this model is a table.") }}
+          {{ adapter.drop_relation(old_relation) }}
+      {% endif %}
+      {# -- now create or replace the table because we're in full-refresh #}
+      {{create_table_as(false, target_relation, source_sql)}}
+    {%- endcall -%}
+
+  {%- else -%}
+    {# -- here is the incremental part #}
+    {% set dest_columns = adapter.get_columns_in_relation(target_relation) %}
+    {% set dest_cols_csv = dest_columns | map(attribute='quoted') | join(', ') %}
+    {%- call statement('main') -%}
+    
+      {%- if unique_key is none -%}
+        {# -- if no unique_key is provided run regular insert as Snowflake may complain #}
+        insert into {{ target_relation }} ({{ dest_cols_csv }})
```

Comment:
this is a really good fix for the `on false` issue with Snowflake's merge statements. Do you think it makes sense to put this logic here? Or should we move it into the Snowflake implementation of `get_merge_sql`?

I like the idea of making materializations represent business logic instead of database logic, as they become a lot more generalizable. Curious what you think!

Reply:
I think that makes total sense! I actually was feeling a bit "awkward" about having this logic sit there but didn't think too much about where else it could live and this is very good, so I'm going to go ahead and change this as you suggest.

Reply:
Great! I think [this](https://github.com/fishtown-analytics/dbt/blob/dev/wilt-chamberlain/plugins/snowflake/dbt/include/snowflake/macros/materializations/merge.sql) would be the place to implement it. If `unique_key` is provided, then we can proceed with `common_get_merge_sql`, otherwise we should return the `insert` statement you've built here

Reply:
yep, its exactly what I just started doing!

Reply:
One thing, I realised there is no incremental deletes anymore, and the merge statement doesn't call a delete. Would you think we need it here?

Reply:
The previous implementation of `incremental` models on Snowflake used `delete` statements to approximate an upsert. Before we did:
```
create temp table as (select * from model code)
delete from destination table where destination unique key = temp table unique key
insert into destination table (select * from temp table)
```

So, records were only deleted _if they were going to be immediately re-inserted_. We'd actually prefer not to call a `delete`, and instead use the `merge` to update these rows in-place. This should be handled by the `when matched` clause in the `merge` statement.

I do think there's a conversation to be had about performance. I wonder if there's any difference between:
1. Deleting existing records and reinserting them (with new values)
2. Updating existing records in place

### An example

**Destination table**

| unique_key | value |
| --- | --- |
| 1 | abc |
| 2 | def |

**Temp table (generated from model `select`)**

| unique_key | value |
| --- | --- |
| 2 | ghi |
| 3 | xyz |

**Desired destination table state**

| unique_key | value |
| --- | --- |
| 1 | abc |
| 2 | ghi |
| 3 | xyz |

So, there are two ways to accomplish this desired end-state. We can either (pseudocode):

**1. delete + insert**
```
delete from destination table where id = 2
insert into destination table (id, value) values (2, ghi), (3, xyz)
```

**2. update + insert (via `merge`)**
```
merge into destination table
from temp table
when matched update -- updates row with id = 2
when not matched insert -- adds rows with id = 3
```

This does raise an interesting question about edge-case behavior with `merge`. What happens if there are duplicate `unique_id`s in either 1) the destination table or 2) the staging table? 

Previously, it was straightforward to understand how the `delete` + `insert` pattern behaved. While having a duplicated `unique_key` would probably lead to undesirable results, the `insert` and `delete` queries would execute successfully.

With the `merge` implementation, I think users will see an error about [non-deterministic results](https://docs.snowflake.net/manuals/sql-reference/sql/merge.html#duplicate-join-behavior) if their `unique_key` is not unique! All told, I think this will actually be a good thing, as it should help alert users to bugs in their model code.

Reply:
Good catch. From what you say here's what I think. Merge is definitely the preferable option and I think unless there's really a good reason for it, you should be getting an error if you're trying to insert dupes. There is probably something fucked up with the source.

Alternatively we could add support for the `ERROR_ON_NONDETERMINISTIC_MERGE` session parameter (when `FALSE` it would pick one of the duplicated rows and insert it) but there doesn't seem to be a clear way on how to select the row and I think this is just bad anyway. I don't really see the point of inserting a dupe row. So I agree with your last point in that comment. So I think the current implementation is cool.